packages/dev/core/test/unit/DeviceInput/babylon.deviceInput.test.ts

```diff
@@ -0,0 +1,226 @@
+        const mouseSource = new DeviceSource(deviceInputSystem, DeviceType.Mouse, 0);
+        onInputChanged = (deviceType: DeviceType, deviceSlot: number, eventData: IUIEvent) => {
+            mouseSource.onInputChangedObservable.notifyObservers(eventData as IPointerEvent);
+        };
+        mouseSource.onInputChangedObservable.add((eventData) => {
+            resultEvent = eventData;
+        });
+
+        expect(mouseSource.onInputChangedObservable.hasObservers()).toBe(true);
+        const mouseEvent: IUIEvent = { pointerId: 0, type: "pointerdown", pointerType: "mouse", clientX: 1, clientY: 1 } as IPointerEvent;
+        onInputChanged(DeviceType.Mouse, 0, mouseEvent);
+
+        expect(resultEvent).toEqual(mouseEvent);
+    });
+});
+
+describe("DeviceSourceManager", () => {
+    let engine: Nullable<NullEngine> = null;
+
+    beforeEach(function () {
+        engine = new NullEngine();
+        engine._deviceSourceManager = new MockInternalDeviceSourceManager() as InternalDeviceSourceManager;
+    });
+
+    afterEach(function () {
+        engine?.dispose();
+    });
+
+    it("should exist", () => {
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        expect(deviceSourceManager).not.toBe(null);
+        expect(engine?._deviceSourceManager?._refCount).toBe(1);
+    });
+
+    it("can use getDeviceSource", () => {
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const touchSource = new DeviceSource(deviceInputSystem, DeviceType.Touch, 0);
+        const touchSource2 = new DeviceSource(deviceInputSystem, DeviceType.Touch, 1);
+        const nullSource = deviceSourceManager.getDeviceSource(DeviceType.Touch, 0);
+        expect(nullSource).toBe(null);
+
+        deviceSourceManager._addDevice(touchSource);
+        deviceSourceManager._addDevice(touchSource2);
+
+        const firstAvailableSource = deviceSourceManager.getDeviceSource(DeviceType.Touch);
+        const specificSource = deviceSourceManager.getDeviceSource(DeviceType.Touch, 1);
+
+        expect(firstAvailableSource).toEqual(touchSource);
+        expect(specificSource).toEqual(touchSource2);
+
+        deviceSourceManager._removeDevice(DeviceType.Touch, 0);
+
+        const nextFirstSource = deviceSourceManager.getDeviceSource(DeviceType.Touch);
+        expect(nextFirstSource).toEqual(touchSource2);
+    });
+
+    it("can use getDeviceSources", () => {
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const touchSource = new DeviceSource(deviceInputSystem, DeviceType.Touch, 0);
+        const touchSource2 = new DeviceSource(deviceInputSystem, DeviceType.Touch, 1);
+        const emptyArray = deviceSourceManager.getDeviceSources(DeviceType.Touch);
+
+        expect(emptyArray.length).toBe(0);

```

Comment:
nit: Maybe group these together so that it's easier to read

Reply:
I'll clean up this section as well